### PR TITLE
forbid the old token format

### PR DESCRIPTION
**Pull request by stlaz**
let's watch the world burn

### CODE REVIEW

**Comment by sttts** [comment]
@stlaz how can we make sure we don't break customers? We need a release-note for sure. But we probably also want some kind of alert in 4.7 ? cc @bergerhoffer

**Comment by bergerhoffer** [comment]
> @stlaz how can we make sure we don't break customers? We need a release-note for sure. But we probably also want some kind of alert in 4.7 ? cc @bergerhoffer

@sttts If we're disallowing the old format tokens starting with 4.8 then yeah I think we'd need to mention something in the docs earlier than 4.8 as an FYI for customers. I don't know whether a z-stream note wold be sufficient, but we can check. An alert also sounds like a good idea so that people aren't surprised.

**Comment by EmilyM1** [comment]
This ok to restest?

**Comment by stlaz** [comment]
> This ok to restest?

until https://github.com/openshift/origin/pull/25965 merges, the e2e tests are likely to fail

**Comment by stlaz** [comment]
/retest
the origin PR merged, let's see what else we've got

**Comment by stlaz** [comment]
/test e2e-aws-serial
Looks like there are no token-related failures in the serial test suite, rerunning to make sure

**Comment by stlaz** [comment]
https://github.com/openshift/origin/pull/26009 should fix all the remaining global-e2e tests issues

**Comment by stlaz** [comment]
/retest
the origin PR merged

**Comment by stlaz** [comment]
/retest
unrelated

File: pkg/oauth/apis/oauth/validation/validation.go [hunk 1]
```diff
@@ -45,6 +46,10 @@ func ValidateTokenName(name string, prefix bool) []string {
 	if len(name) < MinTokenLength {
 		return []string{fmt.Sprintf("must be at least %d characters long", MinTokenLength)}
 	}
+
+	if !strings.HasPrefix(name, sha256Prefix) {
+		return []string{fmt.Sprintf("only the new format of tokens (prefixed with sha256~) is allowed")}
+	}
```
Comment (sttts) [inline pkg/oauth/apis/oauth/validation/validation.go:52]: will this update of existing tokens?

Comment (stlaz) [reply pkg/oauth/apis/oauth/validation/validation.go:52]: Current validation does not allow updating anything else but the `inactivityTimeoutSeconds`, so existing tokens will not be updated. The `inactivityTimeoutSeconds` of existing tokens will no longer be possible to be updated, so those will just time out.

Other, long-lived pre-existing tokens may still survive.

Comment (sttts) [reply pkg/oauth/apis/oauth/validation/validation.go:52]: ack

File: pkg/tokenvalidation/bootstrapauthenticator.go [hunk 2]
```diff
@@ -37,14 +38,14 @@ func NewBootstrapAuthenticator(tokens oauthclient.OAuthAccessTokenInterface, get
...
 
 func (a *bootstrapAuthenticator) AuthenticateToken(ctx context.Context, name string) (*kauthenticator.Response, bool, error) {
-	// hash token for new-style sha256~ prefixed token
-	// TODO: reject non-sha256 prefix tokens in 4.7+
-	if strings.HasPrefix(name, sha256Prefix) {
-		withoutPrefix := strings.TrimPrefix(name, sha256Prefix)
-		h := sha256.Sum256([]byte(withoutPrefix))
-		name = sha256Prefix + base64.RawURLEncoding.EncodeToString(h[0:])
+	if !strings.HasPrefix(name, sha256Prefix) {
+		errOldFormat = errors.New("old token format")
```
Comment (sttts) [inline pkg/tokenvalidation/bootstrapauthenticator.go:42]: `return nil, false, errOldFormat` ?

File: pkg/oauth/apis/oauth/validation/validation.go (hunk 1 above)
Comment (stlaz) [reply pkg/oauth/apis/oauth/validation/validation.go:52]: btw: the newest changes make the old tokens disfunctional for authn

File: pkg/tokenvalidation/tokenauthenticator.go [hunk 3]
```diff
@@ -17,7 +17,10 @@ import (
 	userclient "github.com/openshift/client-go/user/clientset/versioned/typed/user/v1"
 )
 
-var errLookup = errors.New("token lookup failed")
+var (
+	errLookup    = errors.New("token lookup failed")
+	errOldFormat = errors.New("old token format")
```
Comment (sttts) [inline pkg/tokenvalidation/tokenauthenticator.go:22]: would rephrase this like "old, insecure token format" ?

File: pkg/oauth/apis/oauth/validation/validation.go [hunk 4]
```diff
@@ -45,6 +46,10 @@ func ValidateTokenName(name string, prefix bool) []string {
 	if len(name) < MinTokenLength {
 		return []string{fmt.Sprintf("must be at least %d characters long", MinTokenLength)}
 	}
+
+	if !strings.HasPrefix(name, sha256Prefix) {
+		return []string{fmt.Sprintf("only the new format of tokens (prefixed with sha256~) is allowed")}
```
Comment (mfojtik) [inline pkg/oauth/apis/oauth/validation/validation.go:51]: drop `fmt.Sprintf` and just use string
